File: cpp/3623.cpp
class Solution {
public:
    int countTrapezoids(vector<vector<int>>& points) {
        unordered_map<int, int> yCounts;
        yCounts.reserve(points.size());
        for(auto &point : points)
            yCounts[point[1]]++;
        
        long long sum = 0;
        long long sumSquares = 0;
        
        for(auto &[y, count] : yCounts) {
            if(count <= 1) continue;
            const long long pairs = (long long)count * (count - 1) / 2;
            sum += pairs;
            sumSquares += pairs * pairs;
        }
        
        // Use formula: Σ_{i<j} a_i * a_j = ((Σ a_i)² - Σ a_i²) / 2
        long long result = sum * sum - sumSquares;
        result /= 2;
        result %= MOD;
        
        return (int)result;
    }
private:
    static const int MOD = 1e9 + 7;
};

File: python/3623.py
from typing import List


class Solution:
    def countTrapezoids(self, points: List[List[int]]) -> int:
        yCounts = {}
        for point in points:
            yCounts[point[1]] = yCounts.get(point[1], 0) + 1
        sum = 0
        sumSquares = 0
        for _, count in yCounts.items():
            if count <= 1:
                continue
            c = count
            pairs = c * (c - 1) // 2
            sum += pairs
            sumSquares += pairs * pairs
        return (sum * sum - sumSquares) // 2 % (10**9 + 7)